src/fidesops/task/refine_target_path.py

```diff
@@ -0,0 +1,147 @@
+    the path will be expanded to include indices where applicable.
+
+    :param row: Record retrieved from a dataset
+    :param target_path: A list of strings representing the path to the desired field.
+    :param only: A list of values that were used to build the query.
+    :return: A list or a list of lists containing more detailed path(s) to the data in "only".  If there
+    was just one path, we return one list.
+
+    :Example:
+    In this example, path ["A", "B", "C"] points to two elements that match values "F" or "G".  We update
+    the path to insert the indices to locate the appropriate value.
+
+    refine_target_path({"A": {"B": [{"C": "D"}, {"C": "F"}, {"C": "G"}]}}, ["A", "B", "C"], only=["F", "G"])
+
+    [["A", "B", 1, "C"], ["A", "B", 2, "C"]]
+    """
+    try:
+        current_level = target_path[0]
+        current_elem = row[current_level]
+    except KeyError:  # FieldPath not found in record, this is expected to happen when data doesn't exist in collection
+        return []
+    except (IndexError, TypeError):  # No field path or invalid field path
+        logger.warning(f"Error with locating {target_path} on row")
+        return []
+
+    if isinstance(current_elem, dict):
+        next_levels = refine_target_path(current_elem, target_path[1:], only)
+        return _update_path(current_level, next_levels)
+
+    if isinstance(current_elem, list):
+        next_levels = _enter_array(current_elem, target_path[1:], only)
+        return _update_path(current_level, next_levels)
+
+    # Simple case - value is a scalar
+    return [current_level] if _match_found(current_elem, only) else []
+
+
+def _enter_array(array: List[Any], field_path: List[str], only: List[Any]) -> List:
+    """
+    Used by recursive "refine_target_path" whenever arrays are encountered in the row.
+    """
+    results: List[Any] = []
+    for index, elem in enumerate(array):
+        current_result = []
+
+        if field_path:
+            next_result = refine_target_path(elem, field_path, only)
+            current_result = _update_path(index, next_result)
+        else:
+            if isinstance(elem, list):
+                next_result = _enter_array(
+                    elem, field_path, only
+                )  # Nested enter_array calls needed for lists in lists
+                current_result = _update_path(index, next_result)
+            else:
+                if _match_found(elem, only):
+                    current_result = [index]
+
+        if current_result:  # Match found at lower level
+            if isinstance(current_result[0], list):
+                # Keeps nested lists at most list of lists
+                results.extend(current_result)
+            else:
+                results.append(current_result)
+
+    return results[0] if len(results) == 1 else results
+
+
+def _match_found(elem: Any, only: List[Any]) -> bool:
+    """The given scalar element matches one of the input values"""
+    return elem in only
```

Comment:
A corner case surfaced in testing today related to type coercion, and the cons of doing this filtering in code instead of at the database level (although I still think there are a lot of good reasons to do it in code).

Say the query was `select * from customer where id = '1'` as a string, but the field is an integer field.  In several cases, the database language is capable of returning the record where id is 1, even though it was queried as a string.  If data of one type is being used to find data of another type (and the database language was able to make that conversion), our equality filters here will return False because the data is not the same type - the data is 1, and python is looking for string `1`.

Thinking more on this today..

Reply:
addressed here https://github.com/ethyca/fidesops/pull/194/commits/741c9dd34a2e5ed43e2333eb5ae3fb507f58dc57